### DISCUSSION TITLE
Need help with querying for data containing nested JSON objects.

### DISCUSSION THREAD

**Discussion by anuj-gupta-nw** [sqlalchemy/sqlalchemy#9057] [Usage Questions]
Hello,

I need help with querying for data containing nested JSON objects.
My class looks like this -

```python
import yaml
from hermes.sql.types import Json, Uuid
from sqlalchemy import Column

class ChildModel(BaseModel):
  child_uuid = Column(Uuid, primary_key=True)

  child_info = Column(Json(schema=yaml.safe_load(
    """
    type: object
    additionalProperties: false
    properties:
      toy_list:
        description: List of toys.
        type: array
        items:
          $ref: "#/definitions/toy_info"
    definitions:
      toy_info:
        type: object
        additionalProperties: false
        required:
          - toy_uuid
        properties:
          toy_uuid:
            type: string
          feature_list:
            type: array
            items:
              type: string
    """)))

``` 

An example of an entry (row) in my MySQL database -

Column child_uuid -
`a2086e9c-2c44-41c0-8ec5-3374c5693e6f`

Column child_info -
```
{
   "toy_list": [
         {
             "toy_uuid": "9a3d60cc-0e04-4e1a-b23a-618cc66ec358", 
             "feature_list": ["walk", "jump"]
         },
         {
            "toy_uuid": "e258a940-5c47-41db-a9fb-338654788bc6",
            "feature_list": ["crawl"]
         }
   ]
}
```

**I need a query which returns all entries which have "toy_uuid" == a particular UUID.
What should the query look like to achieve this ?**

I have tried my best reading through sqlalchemy documentations and googling but could not figure it out.
Closest I could reach was referring to [this](https://stackoverflow.com/questions/38751438/querying-value-from-json-field-in-flask-sqlalchemy-with-postgresql).

So I started with the following subquery -
```
toys_subquery = session.query(func.json_array_elements(ChildModel.child_info['toy_list']).label('toys')).subquery()
```

but this is failing with error -
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/operators.py", line 440, in __getitem__
    return self.operate(getitem, index)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py", line 317, in operate
    return op(self.comparator, *other, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/operators.py", line 440, in __getitem__
    return self.operate(getitem, index)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/properties.py", line 423, in operate
    return op(self.__clause_element__(), *other, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/operators.py", line 440, in __getitem__
    return self.operate(getitem, index)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 869, in operate
    return op(self.comparator, *other, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/operators.py", line 440, in __getitem__
    return self.operate(getitem, index)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/type_api.py", line 1366, in operate
    return super(TypeDecorator.Comparator, self).operate(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/type_api.py", line 77, in operate
    return o[0](self.expr, op, *(other + o[1:]), **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/default_comparator.py", line 181, in _getitem_impl
    _unsupported_impl(expr, op, other, **kw)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/default_comparator.py", line 185, in _unsupported_impl
    raise NotImplementedError(
NotImplementedError: Operator 'getitem' is not supported on this expression
```

Please help me figure out the final query.

**Comment by zzzeek** [comment]
this is an issue with the `hermes.sql.types` library you're using, you should consult their documentation for how to use SQL operators to access JSON array elements.    The code is correct for SQLAlchemy's JSON type, as illustrated below:

```py
from sqlalchemy import Column
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import JSON
from sqlalchemy import select
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class ChildModel(Base):
    __tablename__ = "child"
    child_uuid = Column(Integer, primary_key=True)

    child_info = Column(JSON())


s = Session()

toys_subquery = s.query(
    func.json_array_elements(ChildModel.child_info["toy_list"]).label("toys")
).subquery()

print(select(toys_subquery))

```

output:

```
SELECT anon_1.toys 
FROM (SELECT json_array_elements(child.child_info[:child_info_1]) AS toys 
FROM child) AS anon_1

```

**Reply by anuj-gupta-nw** [reply]
Thanks for replying Mike.

The `hermes.sql.types` library has -
```python
import json
import jsonschema.validators
import six
from sqlalchemy.dialects import mysql
from sqlalchemy.types import BigInteger, String, TypeDecorator, UnicodeText
import uuid

class Json(TypeDecorator):
  impl = UnicodeText
  cache_ok = False

  def __init__(self, schema=None):
    super(Json, self).__init__()
    self.with_variant(mysql.LONGTEXT(), "mysql") # 2^32 bytes
    if schema is not None:
      cls = jsonschema.validators.validator_for(schema)
      cls.check_schema(schema)
      self._validate = cls(schema).validate
    else:
      self._validate = lambda x: None

  def process_bind_param(self, value, dialect):
    """
    Serializes a python dictionary as JSON.
    """
    if value is not None:
      self._validate(value)
      return six.text_type(json.dumps(value))

  def process_result_value(self, value, dialect):
    """
    Deserializes a JSON string as a python dictionary.
    """
    if value is not None:
      return json.loads(value)
```

> you should consult their documentation for how to use SQL operators to access JSON array elements.

It seems that the `Json` class is simply serializing and deserializing JSON string.
So I am not sure how else to use SQL operators other than what we did with -
```
toys_subquery = s.query(
    func.json_array_elements(ChildModel.child_info["toy_list"]).label("toys")
).subquery()
```

**Reply by zzzeek** [reply]
so that's not a JSON object to start with.

assuming this is PostgreSQL, you'd need to cast this to JSON so that you could use json_array_elements on it:

```py
from sqlalchemy import JSON, cast

toys_subquery = s.query(
    func.json_array_elements(cast(ChildModel.child_info, JSON)["toy_list"]).label("toys")
).subquery()
```

**Reply by anuj-gupta-nw** [reply]
Sorry I forgot to mention that my DB is MariaDB, not PostgreSQL.
So `json_array_elements` won't work as @CaselIT mentioned.

Also I could not understand why you wrote -
>  so that's not a JSON object to start with.

`process_bind_param` method above serializes the `value` passed to json to write it to db. So why do we need to cast it to sqlalchemy.JSON type to be able to do json operations on it ?

**Reply by CaselIT** [reply]
> Sorry I forgot to mention that my DB is MariaDB, not PostgreSQL. So `json_array_elements` won't work as @CaselIT mentioned.
> 
> Also I could not understand why you wrote -
> 
> > so that's not a JSON object to start with.
> 
> `process_bind_param` method above serializes the `value` passed to json to write it to db. So why do we need to cast it to sqlalchemy.JSON type to be able to do json operations on it ?

what mike mentioned is that it's not using a sqlalchemy json type in the type decorator. The `[]` operator is implemented only for some types, since in general you doing `some_column[something]` has no meaning. That's the error you are receiving there.
To support it the type you are using should be changed, either by using the `JSON` type from sqlalchemy or by the hermes library authors

**Comment by CaselIT** [comment]
Looking a bit I don't think json_array_elements exists in mysql. You should first start by trying to create a sql query that does what you want. Once you have it, if needed, we can help creating a Sqlalchemy version that does the same

Also as mike mentioned the type you are using does not seem to be support json operations

**Reply by anuj-gupta-nw** [reply]
Thanks for pointing that out. My db is MariaDb, so `json_array_elements` won't work.
I am not proficient with sql query, but I will try to figure it out.
